src/Microsoft.ML.Transforms/Text/LdaTransform.cs

```diff
@@ -169,224 +176,668 @@ private sealed class ColInfoEx
+                        // REVIEW: This will work, but there are opportunities for caching
+                        // based on input.Counter that are probably worthwhile given how long inference takes.
+                        getSrc(ref src);
+                        lda.Output(in src, ref dst, numBurninIter, reset);
+                    };
+            }
+        }
+
+        public const string LoaderSignature = "LdaTransform";
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "LIGHTLDA",
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature,
+                loaderAssemblyName: typeof(LdaTransform).Assembly.FullName);
+        }
+
+        private readonly TransformInfo[] _exes;
+        private readonly LdaState[] _ldas;
+        private readonly ColumnType[] _types;
+
+        private const string RegistrationName = "LightLda";
+        private const string WordTopicModelFilename = "word_topic_summary.txt";
+        internal const string Summary = "The LDA transform implements LightLDA, a state-of-the-art implementation of Latent Dirichlet Allocation.";
+        internal const string UserName = "Latent Dirichlet Allocation Transform";
+        internal const string ShortName = "LightLda";
+
+        private static (string input, string output)[] GetColumnPairs(ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(columns, nameof(columns));
+            return columns.Select(x => (x.Input, x.Output)).ToArray();
+        }
+
+        internal LdaTransform(IHostEnvironment env, IDataView input, ColumnInfo[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(LdaTransform)), GetColumnPairs(columns))
+        {
+            _exes = new TransformInfo[columns.Length];
+            _types = new ColumnType[columns.Length];
+            _ldas = new LdaState[columns.Length];
+
+            for (int i = 0; i < columns.Length; i++)
+            {
+                var ex = new TransformInfo(Host, columns[i]);
+                _exes[i] = ex;
+                _types[i] = new VectorType(NumberType.Float, ex.NumTopic);
+            }
+            using (var ch = Host.Start("Train"))
+            {
+                Train(ch, input, _ldas);
+            }
+        }
+
+        private LdaTransform(IHost host, ModelLoadContext ctx) : base(host, ctx)
+        {
+            Host.AssertValue(ctx);
+
+            // *** Binary format ***
+            // <prefix handled in static Create method>
+            // <base>
+            // ldaState[num infos]: The LDA parameters
+
+            // Note: columnsLength would be just one in most cases.
+            var columnsLength = ColumnPairs.Length;
+            _exes = new TransformInfo[columnsLength];
+            _ldas = new LdaState[columnsLength];
```

Comment:
>LdaState [](start = 24, length = 8)

are these guys reentrant?

Reply:
could you kindly clarify what you mean by 're-entrant' ?

---
In reply to: [232437876](https://github.com/dotnet/machinelearning/pull/1410#discussion_r232437876) [](ancestors = 232437876)

Reply:
re-entrant means that any methods can be called in parallel from multiple threads

---
In reply to: [232470000](https://github.com/dotnet/machinelearning/pull/1410#discussion_r232470000) [](ancestors = 232470000,232437876)